app/services/payment_providers/moneyhash_service.rb

```diff
@@ -0,0 +1,159 @@
+# frozen_string_literal: true
+
+module PaymentProviders
+  class MoneyhashService < BaseService
+    INTENT_WEBHOOKS_EVENTS = %w[intent.processed intent.time_expired].freeze
+    TRANSACTION_WEBHOOKS_EVENTS = %w[transaction.purchase.failed transaction.purchase.pending_authentication transaction.purchase.successful].freeze
+    CARD_WEBHOOKS_EVENTS = %w[card_token.created card_token.updated card_token.deleted].freeze
+    ALLOWED_WEBHOOK_EVENTS = (INTENT_WEBHOOKS_EVENTS + TRANSACTION_WEBHOOKS_EVENTS + CARD_WEBHOOKS_EVENTS).freeze
+
+    PAYMENT_SERVICE_CLASS_MAP = {
+      "Invoice" => Invoices::Payments::MoneyhashService,
+      "PaymentRequest" => PaymentRequests::Payments::MoneyhashService
+    }.freeze
+
+    def create_or_update(**args)
+      payment_provider_result = PaymentProviders::FindService.call(
+        organization_id: args[:organization].id,
+        code: args[:code],
+        id: args[:id],
+        payment_provider_type: "moneyhash"
+      )
+
+      moneyhash_provider = if payment_provider_result.success?
+        payment_provider_result.payment_provider
+      else
+        PaymentProviders::MoneyhashProvider.new(
+          organization_id: args[:organization].id,
+          code: args[:code]
+        )
+      end
+
+      moneyhash_provider.api_key
+      old_code = moneyhash_provider.code
+      moneyhash_provider.api_key = args[:api_key] if args.key?(:api_key)
+      moneyhash_provider.code = args[:code] if args.key?(:code)
+      moneyhash_provider.name = args[:name] if args.key?(:name)
+      moneyhash_provider.flow_id = args[:flow_id] if args.key?(:flow_id)
+
+      moneyhash_provider.save(validate: false)
+
+      if payment_provider_code_changed?(moneyhash_provider, old_code, args)
+        moneyhash_provider.customers.update_all(payment_provider_code: args[:code]) # rubocop:disable Rails/SkipsModelValidations
+      end
+
```

Comment:
Make API call to get the signature key from moneyhash. You can make it synchronous so we can return an error if something goes wrong or in a job if you prefer.
You can save the key in the payment_provider settings 👍